cdap-app-fabric/src/main/java/io/cdap/cdap/app/preview/DefaultPreviewRunnerManager.java

```diff
@@ -0,0 +1,216 @@
+
+  @Inject
+  DefaultPreviewRunnerManager(
+    @Named(PreviewConfigModule.PREVIEW_CCONF) CConfiguration previewCConf,
+    @Named(PreviewConfigModule.PREVIEW_HCONF) Configuration previewHConf,
+    @Named(PreviewConfigModule.PREVIEW_SCONF) SConfiguration previewSConf,
+    SecureStore secureStore, DiscoveryService discoveryService,
+    @Named(DataSetsModules.BASE_DATASET_FRAMEWORK) DatasetFramework datasetFramework,
+    TransactionSystemClient transactionSystemClient,
+    @Named(PreviewConfigModule.PREVIEW_LEVEL_DB) LevelDBTableService previewLevelDBTableService,
+    PreviewRunnerModule previewRunnerModule, PreviewRequestFetcherFactory previewRequestFetcherFactory) {
+    this.previewCConf = previewCConf;
+    this.previewHConf = previewHConf;
+    this.previewSConf = previewSConf;
+    this.datasetFramework = datasetFramework;
+    this.secureStore = secureStore;
+    this.discoveryService = discoveryService;
+    this.transactionSystemClient = transactionSystemClient;
+    this.maxConcurrentPreviews = previewCConf.getInt(Constants.Preview.CACHE_SIZE, 10);
+    this.previewPollers = new ConcurrentHashMap<>();
+    this.previewRunnerModule = previewRunnerModule;
+    this.previewLevelDBTableService = previewLevelDBTableService;
+    this.previewRequestFetcherFactory = previewRequestFetcherFactory;
+  }
+
+  @Override
+  protected synchronized void startUp() throws Exception {
+    previewInjector = createPreviewInjector();
+    // Create and start the preview poller services.
+    for (int i = 0; i < maxConcurrentPreviews; i++) {
+      String pollerInfo = UUID.randomUUID().toString();
+
+      PreviewRunnerService pollerService = new PreviewRunnerService(
+        previewCConf, previewInjector.getInstance(PreviewRunner.class),
+        previewRequestFetcherFactory.create(Bytes.toBytes(pollerInfo)));
+
+      pollerService.startAndWait();
+      previewPollers.put(pollerInfo, pollerService);
+    }
+    PreviewRunner runner = previewInjector.getInstance(PreviewRunner.class);
+    if (runner instanceof Service) {
+      ((Service) runner).startAndWait();
+    }
+  }
+
+  @Override
+  protected synchronized void shutDown() throws Exception {
+    PreviewRunner runner = previewInjector.getInstance(PreviewRunner.class);
+    if (runner instanceof Service) {
+      stopQuietly((Service) runner);
+    }
+
+    for (Service pollerService : previewPollers.values()) {
+      stopQuietly(pollerService);
+    }
+  }
+
+  private void stopQuietly(Service service) {
+    try {
+      service.stopAndWait();
+    } catch (Exception e) {
+      LOG.debug("Error stopping the preview runner.", e);
+    }
+  }
+
+  @Override
+  public void stop(byte[] runnerId) throws Exception {
+    PreviewRunnerService service = previewPollers.get(Bytes.toString(runnerId));
+    if (service == null) {
+      String msg = "Preview run cannot be stopped. Please try stopping again or start new preview run.";
```

Comment:
No need to declare `msg`. Just move the message into the Exception constructor.

Reply:
done.